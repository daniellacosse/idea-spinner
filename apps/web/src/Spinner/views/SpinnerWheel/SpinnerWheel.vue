<script lang="ts" setup>
  import { ref, watch } from "vue";

  // see: https://github.com/vuejs/core/issues/4294
  interface SpinnerWheelProps {
    description?: string;
    isLocked: boolean;
    isSpinning: boolean;
    name: string;
    value?: string;
  }

  const props = defineProps<SpinnerWheelProps>();

  const isAnimating = ref(false);
  const input = ref<HTMLInputElement | null>(null);
  const displayEmptyValue = "--";

  watch(
    () => props.value,
    () => {
      isAnimating.value = true;
    }
  );

  const emit = defineEmits(["edit", "lock", "unlock"]);

  function clearChange() {
    isAnimating.value = false;
  }

  function handleEdit({ target }: Event) {
    const { value } = target as HTMLInputElement;

    emit("edit", { value });
  }
</script>

<template>
  <li
    :class="{
      SpinnerWheel: true,
      'SpinnerWheel--stopped': isLocked || !isSpinning,
      'SpinnerWheel--tick': !isLocked && isAnimating && isSpinning,
    }"
    @animationend="clearChange"
  >
    <header class="SpinnerWheel__header">
      <h2 class="SpinnerWheel__name">{{ name }}</h2>
      <i
        :class="{
          SpinnerWheel__lock: true,
          'SpinnerWheel__lock--locked': isLocked,
        }"
        role="button"
        @click.prevent="
          !isSpinning && $emit(isLocked ? 'unlock' : 'lock', name)
        "
      >
        {{ isLocked ? "🔒" : "🔓" }}
      </i>
    </header>
    <button
      :class="{
        SpinnerWheel__value: true,
        'SpinnerWheel__value--description':
          Boolean(description) && (isLocked || !isSpinning),
      }"
      :title="description"
      @click.stop.prevent="input?.focus()"
    >
      <!--
        ISSUE #32: somehow get contenteditable to work

        For the life of me I could not get contenteditable 
        to work in Vue3 - so, my clever workaround was to bind the 
        size attribute to the current input value's length.
      -->
      <input
        class="SpinnerWheel__valueInput"
        :size="value?.length ?? displayEmptyValue.length"
        :value="value ?? displayEmptyValue"
        :disabled="value === undefined || isSpinning"
        @input="handleEdit"
        ref="input"
      />
    </button>
  </li>
</template>

<style scoped>
  * {
    all: initial;
  }

  .SpinnerWheel {
    animation-duration: var(--animation-timing);
    background: var(--color-primary);
    border-radius: var(--gutter-narrow);
    flex-direction: column;
    height: 100%;
    overflow: hidden;
    position: relative;
    transition: background var(--animation-timing) ease-out;
  }

  .SpinnerWheel__header {
    background: var(--color-primary-variant);
    border-bottom: var(--hairline) solid var(--color-primary);
    height: var(--gutter-wide);
    line-height: var(--gutter-wide);
    width: 100%;
  }

  .SpinnerWheel__header * {
    user-select: none;
  }

  .SpinnerWheel__name {
    color: var(--color-primary-variant-foreground);
    font-family: var(--font-primary);
    font-size: var(--gutter-narrow);
    font-weight: bold;
    text-transform: uppercase;
  }

  .SpinnerWheel__lock {
    cursor: pointer;
    float: right;
    line-height: var(--gutter-wide);
    opacity: 0.5;
    position: absolute;
    right: var(--gutter-narrow);
  }

  .SpinnerWheel__lock.SpinnerWheel__lock--locked {
    opacity: 1;
  }

  .SpinnerWheel__value {
    box-sizing: border-box;
    flex-grow: 1;
    padding: var(--gutter-narrow);
    width: 100%;
  }

  .SpinnerWheel__valueInput {
    color: var(--color-primary-foreground);
    font-family: var(--font-secondary);
    font-size: var(--gutter-wide);
    max-width: 100%;
    text-align: center;
    text-overflow: ellipsis;
    transition: color var(--animation-timing) ease-out;
  }

  .SpinnerWheel__value--description {
    cursor: help;
  }

  .SpinnerWheel__value--description::before {
    content: "ℹ️";
    padding: var(--gutter-narrow);
  }

  .SpinnerWheel,
  .SpinnerWheel__header,
  .SpinnerWheel__value {
    align-items: center;
    display: flex;
    justify-content: center;
  }

  .SpinnerWheel--tick {
    animation-name: tick;
  }

  @keyframes tick {
    from {
      background-color: var(--color-primary-variant);
      border-color: var(--color-primary-variant-foreground);
      color: var(--color-primary-variant-foreground);
    }

    to {
      background-color: var(--color-primary);
      border-color: var(--color-primary-foreground);
      color: var(--color-primary-foreground);
    }
  }

  .SpinnerWheel--stopped {
    background: var(--color-primary-variant);
  }

  .SpinnerWheel--stopped .SpinnerWheel__valueInput {
    color: var(--color-primary-variant-foreground);
  }

  .SpinnerWheel--stopped > .SpinnerWheel__name {
    border-color: var(--color-primary-variant-foreground);
  }
</style>
