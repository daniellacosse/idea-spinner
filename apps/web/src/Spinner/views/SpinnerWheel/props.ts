export interface SpinnerWheelProps {
  description?: string;
  isSpinning: boolean;
  name: string;
  tickDuration: string;
  value?: string;
}

export default {
  description:
    "Clowns are a curious breed who sustain themselves " +
    "on the laugher of children.",
  isSpinning: false,
  name: "Demo Wheel",
  tickDuration: "200ms",
  value: "Clowns"
} as SpinnerWheelProps;
