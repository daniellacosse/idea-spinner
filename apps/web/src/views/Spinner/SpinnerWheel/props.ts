export interface SpinnerWheelProps {
  isSpinning: boolean;
  name: string;
  value: string;
}

export default {
  isSpinning: false,
  name: "Demo Wheel",
  value: "Clowns"
} as SpinnerWheelProps;
