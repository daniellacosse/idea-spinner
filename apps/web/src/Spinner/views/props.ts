import { SpinnerWheelProps } from "./SpinnerWheel";

export interface SpinnerProps {
  hasSpun: boolean;
  isSpinning: boolean;
  maxColumns?: number;
  wheels: SpinnerWheelProps[];
}

export default {
  hasSpun: false,
  isSpinning: false,
  maxColumns: 4,
  wheels: [{
    isSpinning: true,
    name: "color",
    value: "red",
  }, {
    isSpinning: true,
    name: "shape",
    value: "round",
  }, {
    isSpinning: false,
    name: "feeling",
    value: "sad",
  }]
} as SpinnerProps;
