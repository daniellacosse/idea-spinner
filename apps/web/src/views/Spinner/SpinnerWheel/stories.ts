import SpinnerWheel from "./SpinnerWheel.vue";
import args, { SpinnerWheelProps } from "./props";

export default {
  args,
  component: SpinnerWheel,
  title: "Views/Spinner/Spinner Wheel",
  parameters: {
    previewTabs: {
      "storybook/docs/panel": {
        hidden: true
      }
    }
  }
};

export const Example = (wheel: SpinnerWheelProps) => ({
  components: { SpinnerWheel },
  setup() {
    return { wheel };
  },
  template: "<SpinnerWheel v-bind='wheel' />"
});
