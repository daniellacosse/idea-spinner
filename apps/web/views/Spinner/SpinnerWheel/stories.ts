import SpinnerWheel from "./SpinnerWheel.vue";
import args, { SpinnerWheelProps } from "./props";

export default {
  args,
  component: SpinnerWheel,
  title: "Spinner/Spinner Wheel"
};

export const Example = (wheel: SpinnerWheelProps) => ({
  components: { SpinnerWheel },
  setup() {
    return { wheel };
  },
  template: `<SpinnerWheel v-bind="wheel" />`
});
