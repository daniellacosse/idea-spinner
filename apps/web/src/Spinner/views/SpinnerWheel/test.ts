import props from "./props";
import { shallowMount } from "@vue/test-utils";
import SpinnerWheel from "./SpinnerWheel.vue";
import { expect, test } from "@jest/globals";

test.concurrent("spinner wheel - snapshot", async () => {
  const { element } = shallowMount(SpinnerWheel, { props });

  expect(element).toMatchSnapshot();
});
