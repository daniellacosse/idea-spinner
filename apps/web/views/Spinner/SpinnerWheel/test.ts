import props from "./props";
import { shallowMount } from "@vue/test-utils";
import SpinnerWheel from "./SpinnerWheel.vue";

test.concurrent("spinner wheel - snapshot", async () => {
  const { element } = shallowMount(SpinnerWheel, { props });

  expect(element).toMatchSnapshot();
});
