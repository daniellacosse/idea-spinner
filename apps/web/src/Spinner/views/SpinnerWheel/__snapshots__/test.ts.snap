// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<li
  class="SpinnerWheel SpinnerWheel--stopped"
>
  <header
    class="SpinnerWheel__header"
  >
    <h2
      class="SpinnerWheel__name"
    >
      Demo Wheel
    </h2>
    <i
      class="SpinnerWheel__lock"
      role="button"
    >
      ðŸ”“
    </i>
  </header>
  <button
    class="SpinnerWheel__value SpinnerWheel__value--description"
    title="Clowns are a curious breed who sustain themselves on the laugher of children."
  >
    <!--
        ISSUE #32: somehow get contenteditable to work

        For the life of me I could not get contenteditable to work in Vue3 - so, 
        my clever workaround was to bind the size attribute to the current
        input value's length.
      -->
    <input
      class="SpinnerWheel__valueInput"
      size="6"
    />
  </button>
</li>
`;
